### PR TITLE
Add LoRA support to conditional SFNO

### PR DESCRIPTION
**Pull request by mcgibbon**
Short description of why the PR is needed and how it satisfies those requirements, in sentence form.

Changes:
- Added options `lora_rank`, `lora_alpha`, `spectral_lora_rank`, and `spectral_lora_alpha` controlling the rank and strength of LoRA associated with Conv1x1 operations and with the spectral filter.

- [x] Tests added

### CODE REVIEW

File: fme/core/models/conditional_sfno/lora.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
+from __future__ import annotations
+
+import math
+
+import torch
+import torch.nn as nn
+
+
+class LoRAConv2d(nn.Conv2d):
```
Comment (mcgibbon) [inline fme/core/models/conditional_sfno/lora.py:9]: I don't like that this uses inheritance, but without this it's very difficult to have previously saved checkpoints load properly, or I'd need to copy-paste some lora code every single time we call a Conv2D (which I was doing at first, and it's a lot).

**Comment by mcgibbon** [comment]
This code change does not affect performance of the non-LoRA training.

image

Machine noise is significant, but my baseline runs show if anything better performance on the experimental branch using LoRA. Because I've done nothing to improve performance on that branch, I expect this is noise.

Baseline run forked from main: https://wandb.ai/ai2cm/ace/runs/act5d5bu/overview?nw=ohyfh3fa3lf
Baseline run on experimental branch with LoRA code: https://wandb.ai/ai2cm/ace/runs/udrkj93s/overview?nw=ohyfh3fa3lf

**Comment by mcgibbon** [comment]
Note the LoRA training runs (a few configurations added here) are also within this performance noise envelope:
image

File: fme/core/models/conditional_sfno/sfnonet.py [hunk 2]
```diff
@@ -257,23 +280,32 @@ def __init__(
...
         if use_mlp == True:
-            MLPH = MLP
             mlp_hidden_dim = int(embed_dim * mlp_ratio)
-            self.mlp = MLPH(
+            self.mlp = MLP(
                 in_features=embed_dim,
                 hidden_features=mlp_hidden_dim,
                 act_layer=act_layer,
                 drop_rate=drop_rate,
-                checkpointing=checkpointing,
```
Comment (elynnwu) [inline fme/core/models/conditional_sfno/sfnonet.py:267]: MLP still accepts `checkpointing`, is it removed because we don't use it?  Otherwise we should restore it.

File: fme/core/models/conditional_sfno/s2convolutions.py [hunk 3]
```diff
@@ -67,9 +78,16 @@ def __init__(
         bias=False,
         use_tensorly=True,
         filter_residual: bool = False,
+        lora_rank: int = 0,
+        lora_alpha: float | None = None,
     ):  # pragma: no cover
         super(SpectralConvS2, self).__init__()
 
+        if in_channels != out_channels:
```
Comment (elynnwu) [inline fme/core/models/conditional_sfno/s2convolutions.py:86]: Should this be checked only when LoRA is enabled?

Comment (mcgibbon) [reply fme/core/models/conditional_sfno/s2convolutions.py:86]: This is a hidden limitation of the existing code, not specific to LoRA. Basically it was never used without this config, and a bug was introduced (before us) that broke this case.

File: fme/core/models/conditional_sfno/sfnonet.py (hunk 2 above)
Comment (mcgibbon) [reply fme/core/models/conditional_sfno/sfnonet.py:267]: Good catch, a previous iteration of this PR required this to be removed but it doesn't anymore. Added it back.

We also don't use it, I'd perhaps suggest we remove it. But let's not look into that right now.

File: fme/core/models/conditional_sfno/s2convolutions.py (hunk 3 above)
Comment (mcgibbon) [reply fme/core/models/conditional_sfno/s2convolutions.py:86]: Specifically it's the part in `forward` where `xp = torch.zeros_like(x)` xp is the output tensor, and is initialized to look like the input tensor. This only works if the in and out channels are the same.

I don't think we need that part, it's for some kind of "padding" feature that was already removed/commented out code when we copied this over. But we also don't need in_channels != out_channels, so I disabled that instead.

File: fme/core/models/conditional_sfno/sfnonet.py [hunk 4]
```diff
@@ -195,6 +203,10 @@ def __init__(
         filter_residual=False,
         affine_norms=False,
         filter_num_groups: int = 1,
+        lora_rank: int = 0,
```
Comment (Arcomano1234) [inline fme/core/models/conditional_sfno/sfnonet.py:206]: There should be some documentation somewhere, even if its just so I remember why `lora_rank` vs `spectral_lora_rank` is important (e.g., `spectral_lora_rank` probably should be greater than `lora_rank`)

Comment (mcgibbon) [reply fme/core/models/conditional_sfno/sfnonet.py:206]: The documentation of this and all other arguments is on the Builder class, since that's what gets exposed publicly. I'd rather not keep two copies up to date.

```
        lora_rank: Rank of the LoRA adaptations outside of spectral convolutions.
            0 (default) disables LoRA.
        lora_alpha: Strength of the LoRA adaptations outside of spectral convolutions.
            Defaults to lora_rank.
        spectral_lora_rank: Rank of the LoRA adaptations for spectral convolutions.
            0 (default) disables LoRA.
        spectral_lora_alpha: Strength of the LoRA adaptations for spectral convolutions.
            Defaults to spectral_lora_rank.
```
